sigstore/_internal/timestamping.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2022 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Timestamp Authority Client for `sigstore-python`
+"""
+
+import requests
+from rfc3161_client import (
+    TimestampRequestBuilder,
+    TimeStampResponse,
+    decode_timestamp_response,
+)
+
+from sigstore._internal import USER_AGENT
+
+
+class TimestampError(Exception):
+    """
+    A generic error in the TimestampAuthority client.
+    """
+
+    pass
+
+
+class TimestampAuthorityClient:
+    """Internal client to deal with a Timestamp Authority"""
+
+    def __init__(self, url: str) -> None:
+        """
+        Create a new `TimestampAuthorityClient` from the given URL
+        """
+        self.url = url
+        self.session = requests.Session()
+        self.session.headers.update(
+            {
+                "Content-Type": "application/timestamp-query",
+                "User-Agent": USER_AGENT,
+            }
+        )
+        self.timeout: int = 30
+
+    def __del__(self) -> None:
+        """
+        Terminates the underlying network session.
+        """
+        self.session.close()
+
+    def timestamps(self, signature: bytes) -> TimeStampResponse:
```

Comment:
This one is hard since `timestamp` is both a noun and a verb. Here's one option for a different name:
```suggestion
    def request_timestamp(self, signature: bytes) -> TimeStampResponse:
```